Add data container voter documentation (#1685)
contao/docs: docs/dev/framework/security/data_container.md

---
title: "Data Container permissions"
description: Explanation of creating data container voters.
aliases:
    - /framework/security/data_container/
---

Starting with version 5.0, Contao uses Symfony security voters to check _create_, _read_, _update_ and _delete_ (CRUD) permissions in data containers. This replaces the `checkPermission` onload callbacks that were used in Contao 4 and before. Before implementing your Contao-specific voter, make sure to get familiar with [Symfony security voters](https://symfony.com/doc/current/security/voters.html).

Contao provides an abstract class [AbstractDataContainerVoter](https://github.com/contao/contao/blob/5.3/core-bundle/src/Security/Voter/DataContainer/AbstractDataContainerVoter.php) to simplify voting on CRUD operations for one specific DCA table. See the following example for a typical archive table:

> First, we need the archive dca definition:

Reviewer: ```suggestion
Contao provides an abstract class [AbstractDataContainerVoter](https://github.com/contao/contao/blob/5.3/core-bundle/src/Security/Voter/DataContainer/AbstractDataContainerVoter.php) to simplify voting on CRUD operations for one specific DCA table. See the following example for a typical parent table definition.

First, we need the DCA for the parent table:

```

```php
# contao/dca/tl_example_archive.php

> 

Reviewer: ```suggestion
// contao/dca/tl_example_archive.php
```

use Contao\DC_Table;

$GLOBALS['TL_DCA']['tl_example_archive'] = [
    'config' => [
        'dataContainer' => DC_Table::class,
        'ctable' => ['tl_example_item'],
    ],
    // ...
]
```

> Register the permissions:

Reviewer: ```suggestion
We also register custom permissions. One to define to which parent record the user can have access to and one to define what kind of CRUD permissions are granted for this table:

```

```php
# contao/config/config.php

> 

Reviewer: ```suggestion
// contao/config/config.php
```

$GLOBALS['TL_PERMISSIONS'][] = 'examples';
$GLOBALS['TL_PERMISSIONS'][] = 'examplep';
```

> Add permissions to the user dca (should be also done for user_group!):

Reviewer: ```suggestion
Add permissions to the user DCA (should be also done for user group!):

```

```php
# contao/dca/tl_user.php

> 

Reviewer: ```suggestion
// contao/dca/tl_user.php
```

Contao provides an abstract class [AbstractDataContainerVoter](https://github.com/contao/contao/blob/5.3/core-bundle/src/Security/Voter/DataContainer/AbstractDataContainerVoter.php) to simplify voting on CRUD operations for one specific DCA table. See the following example for a typical parent table definition.

First, we need the DCA for the parent table:

```php
// contao/dca/tl_example_archive.php
use Contao\DC_Table;

$GLOBALS['TL_DCA']['tl_example_archive'] = [
    'config' => [
        'dataContainer' => DC_Table::class,
        'ctable' => ['tl_example_item'],
    ],
    // ...
]
```

We also register custom permissions. One to define to which parent record the user can have access to and one to define what kind of CRUD permissions are granted for this table:

```php
// contao/config/config.php
$GLOBALS['TL_PERMISSIONS'][] = 'examples';
$GLOBALS['TL_PERMISSIONS'][] = 'examplep';
```

> Add permissions to the user DCA (should be also done for user group!):

Reviewer: Btw. shouldn't it primarily be done for the user group? Or do you also always have to do it for the user? I don't actually know 🤔

Author: Typically it is done for both I think. But of course it should be no problem to do it only for groups or users.

---
title: "Data Container permissions"
description: Explanation of creating data container voters.
aliases:
    - /framework/security/data_container/
---

Since Contao 5 data container permissions can be set by data container voters. This replaces the "checkPermission" onload callbacks that were used in Contao 4 and before.

Contao provides an abstract class [AbstractDataContainerVoter](https://github.com/contao/contao/blob/5.3/core-bundle/src/Security/Voter/DataContainer/AbstractDataContainerVoter.php) for data container voters.

> A class extending it is able to vote for data container actions. See the following example for a typical archive table:

Reviewer 2: ```suggestion
Starting with version 5.0, Contao uses Symfony security voters to check _create_, _read_, _update_ and _delete_ (CRUD) permissions in data containers. This replaces the `checkPermission` onload callbacks that were used in Contao 4 and before. Before implementing your Contao-specific voter, make sure to get familiar with [Symfony security voters](https://symfony.com/doc/current/security/voters.html).

Contao provides an abstract class [AbstractDataContainerVoter](https://github.com/contao/contao/blob/5.3/core-bundle/src/Security/Voter/DataContainer/AbstractDataContainerVoter.php) to simplify voting on CRUD operations for one specific DCA table. See the following example for a typical archive table:
```

            $action instanceof DeleteAction => 
                $this->accessDecisionManager->decide($token, ['contao_user.examples'], $action->getCurrentId()) && 
                $this->accessDecisionManager->decide($token, ['contao_user.examplep.delete']),
        };
        // You can also add additional checks/conditions for allowing/disallowing actions here, if your code requires it.
    }
}
```

In your archive list view you need to filter out all items the user has no read access, otherwise you'll get a permission denied exception.
See following example listener that sets the root IDs for the current user:

```php
namespace App\EventListener\DataContainer\ExampleArchive;

use Contao\BackendUser;
use Contao\CoreBundle\DependencyInjection\Attribute\AsCallback;
use Contao\DataContainer;
use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;

#[AsCallback(table: 'tl_example_archive', target: 'config.onload')]
class ConfigOnLoadListener
{
    public function __construct(
        private readonly TokenStorageInterface $tokenStorage,
    )

>     {

Reviewer 2: ```suggestion
    ) {
```

    }

>     public function __invoke(?DataContainer $dc = null): void

Reviewer 2: ```suggestion
    public function __invoke(): void
```

---
title: "Data Container permissions"
description: Explanation of creating data container voters.
aliases:
    - /framework/security/data_container/
---

> Since contao 5 data container permissions can be set by data container voters. This replaces the "checkPermission" onload callbacks that were used in Contao 4 and before.

Reviewer: ```suggestion
Since Contao 5 data container permissions can be set by data container voters. This replaces the "checkPermission" onload callbacks that were used in Contao 4 and before.
```

> Contao comes with an bundles abstract class [AbstractDataContainerVoter](https://github.com/contao/contao/blob/5.3/core-bundle/src/Security/Voter/DataContainer/AbstractDataContainerVoter.php).

Reviewer: Hm, what do you mean by "bundles abstract class" here? Or do you just mean

> Contao provides an abstract class for data container voters.

Author: Typo, should be "bundled"

> A class extending it is able to vote for data container actions. See following example:

Reviewer: ```suggestion
A class extending it is able to vote for data container actions. See the following example:
```

```php
<?php

> 

Reviewer: ```suggestion
```

class ExampleAccessVoter extends AbstractDataContainerVoter
{
    public function __construct(
        private readonly AccessDecisionManagerInterface $accessDecisionManager
    ) {
    }

    protected function getTable(): string
    {
        return ExampleArchiveModel::getTable();
    }

    protected function hasAccess(TokenInterface $token, UpdateAction|CreateAction|ReadAction|DeleteAction $action): bool
    {
        if (!$this->accessDecisionManager->decide($token, ['contao_user.modules.example'])) {
            return false;
        }

        return match (true) {
            $action instanceof CreateAction => $this->accessDecisionManager->decide($token,['contao_user.examplep.create']),
            $action instanceof ReadAction,
            $action instanceof UpdateAction => $this->accessDecisionManager->decide($token,['contao_user.examples'], $action->getCurrentId()),
            $action instanceof DeleteAction => 
                $this->accessDecisionManager->decide($token,['contao_user.examples'], $action->getCurrentId()) && 
                $this->accessDecisionManager->decide($token,['contao_user.examplep.delete']),
        };
    }
}
```

> You still need a onload callback listener to filter out list items not allowed for the user:

Reviewer: ```suggestion
You still need an onload callback listener to filter out list items not allowed for the user:
```

```php
namespace App\EventListener\DataContainer\ExampleArchive;

use Contao\BackendUser;
use Contao\CoreBundle\DependencyInjection\Attribute\AsCallback;
use Contao\DataContainer;

#[AsCallback(table: 'tl_example_archive', target: 'config.onload')]
class ConfigOnLoadListener
{
    public function __invoke(?DataContainer $dc = null): void
    {
        $user = BackendUser::getInstance();

        if ($user->isAdmin) {
            return;

>         }

Reviewer: You always need to use dependency injection. In this case you should inject the [`AuthorizationCheckerInterface`](https://docs.contao.org/5.x/dev/reference/services/#:~:text=AuthorizationCheckerInterface) and check for `$this->authorizationChecker->isGranted('ROLE_ADMIN')`.

Author: Code was taken from the core, but I will change it.

Reviewer: > Code was taken from the core

Likely from a legacy callback listener, not a service?

Author: I did not find any "non-legacy" listener in the core for this ;) 
I adjusted the code to use DI.

Contao provides an abstract class [AbstractDataContainerVoter](https://github.com/contao/contao/blob/5.3/core-bundle/src/Security/Voter/DataContainer/AbstractDataContainerVoter.php) for data container voters.
A class extending it is able to vote for data container actions. See the following example:

```php
namespace App\Security\Voter\DataContainer;

use App\Model\ExampleArchiveModel;
use App\Security\ExamplePermissions;
use Contao\CoreBundle\Security\DataContainer\CreateAction;
use Contao\CoreBundle\Security\DataContainer\DeleteAction;
use Contao\CoreBundle\Security\DataContainer\ReadAction;
use Contao\CoreBundle\Security\DataContainer\UpdateAction;
use Contao\CoreBundle\Security\Voter\DataContainer\AbstractDataContainerVoter;
use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
use Symfony\Component\Security\Core\Authorization\AccessDecisionManagerInterface;

class ExampleAccessVoter extends AbstractDataContainerVoter
{
    public function __construct(
        private readonly AccessDecisionManagerInterface $accessDecisionManager
    ) {
    }

    protected function getTable(): string
    {

>         return ExampleArchiveModel::getTable();

Reviewer 2: I would suggest to use a string here, because models are not necessary and usually not used in the back end. It also does not exist in the code example and might confuse readers.

    }

    protected function hasAccess(TokenInterface $token, UpdateAction|CreateAction|ReadAction|DeleteAction $action): bool
    {

>         if (!$this->accessDecisionManager->decide($token, ['contao_user.modules.example'])) {

Reviewer 2: ```suggestion
        if (!$this->accessDecisionManager->decide($token, [ContaoCorePermissions::USER_CAN_ACCESS_MODULE], 'example')) {
```

I think this is more fool-proof (even though your version works as well). Also needs the `use` statement for `ContaoCorePermissions`.

            return false;
        }

        return match (true) {
            $action instanceof CreateAction => $this->accessDecisionManager->decide($token, ['contao_user.examplep.create']),
            $action instanceof ReadAction,
            $action instanceof UpdateAction => $this->accessDecisionManager->decide($token, ['contao_user.examples'], $action->getCurrentId()),
            $action instanceof DeleteAction => 
                $this->accessDecisionManager->decide($token, ['contao_user.examples'], $action->getCurrentId()) && 
                $this->accessDecisionManager->decide($token, ['contao_user.examplep.delete']),
        };

>     }

Reviewer 2: The docs should also explain what `contao_user.examples` and `contao_user.examplep` are (the "archive" and "permission" flags in the user profile). Maybe with a small DCA, also explaining `TL_PERMISSION`? Or maybe that already exists somewhere?

Author: There is already a doc for this (https://docs.contao.org/5.x/dev/framework/security/#custom-back-end-access-rights), but I added it to the example.

class ExampleAccessVoter extends AbstractDataContainerVoter
{
    public function __construct(
        private readonly AccessDecisionManagerInterface $accessDecisionManager
    ) {
    }

    protected function getTable(): string
    {
        return ExampleArchiveModel::getTable();
    }

    protected function hasAccess(TokenInterface $token, UpdateAction|CreateAction|ReadAction|DeleteAction $action): bool
    {
        if (!$this->accessDecisionManager->decide($token, ['contao_user.modules.example'])) {
            return false;
        }

        return match (true) {
            $action instanceof CreateAction => $this->accessDecisionManager->decide($token, ['contao_user.examplep.create']),
            $action instanceof ReadAction,
            $action instanceof UpdateAction => $this->accessDecisionManager->decide($token, ['contao_user.examples'], $action->getCurrentId()),
            $action instanceof DeleteAction => 
                $this->accessDecisionManager->decide($token, ['contao_user.examples'], $action->getCurrentId()) && 
                $this->accessDecisionManager->decide($token, ['contao_user.examplep.delete']),
        };
    }
}
```

> You still need an onload callback listener to filter out list items not allowed for the user:

Reviewer: What is this actually for? You don't need that for regular listings, do you? Or is this an example for `MODE_TREE`  with some "mounts"?

Author: This is a normal archive view. Otherwise you'll get permission denied in the archives list, when it is not added to the user (group) permissions. Otherwise you do not need the voter in this case.

Reviewer: Hm, I don't really understand. Generally this seems strange to me. The `contao/news-bundle` also does not need this.

Author: Sure it does: https://github.com/contao/contao/blob/33a057e9a1ee2470db8e9094cc36fcef1a108922/news-bundle/contao/dca/tl_news_archive.php#L145
Also events and faq.

Author: To make it complete, we should also add this: https://github.com/contao/contao/blob/33a057e9a1ee2470db8e9094cc36fcef1a108922/news-bundle/contao/dca/tl_news_archive.php#L172

But this is really uncool stuff :)

Reviewer: I see. @aschempp can this really not be implemented purely as a Voter yet? Instead of adjusting the `'root'`s via a callback I mean.

Reviewer 2: A voter decides permissions, it does not decide your view. It is correct that we need to limit the viewable items, because we don't want to load 100'000 records from the database to check permissions on them for no reason.

```php
namespace App\EventListener\DataContainer\ExampleArchive;

use Contao\BackendUser;
use Contao\CoreBundle\DependencyInjection\Attribute\AsCallback;
use Contao\DataContainer;
use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;

#[AsCallback(table: 'tl_example_archive', target: 'config.onload')]
class ConfigOnLoadListener
{
    public function __construct(
        private readonly TokenStorageInterface $tokenStorage,
    )
    {
    }

    public function __invoke(?DataContainer $dc = null): void
    {
        /**
         * @var BackendUser|null $user
         */
        $user = $this->tokenStorage->getToken()?->getUser();

        if ($user && $user->isAdmin) {
            return;
        }

        // Set root IDs
        if (!$user || !\is_array($user->examples) || empty($user->examples)) {
            $root = [0];
        } else {
            $root = $user->examples;

>         }

Reviewer: ```suggestion
        $user = $this->tokenStorage->getToken()?->getUser();

        if (!$user instanceof BackendUser || $user->isAdmin) {
            return;
        }

        // Set root IDs
        if (!$user->examples || !\is_array($user->examples)) {
            $root = [0];
        } else {
            $root = $user->examples;
        }
```
